### PR TITLE
Http Telemetry

### PR DESCRIPTION
**Pull request by bgavrilMS**

### CODE REVIEW

File: src/client/Microsoft.Identity.Client/OAuth2/OAuth2Client.cs [hunk 1]
```diff
@@ -57,6 +57,11 @@ public void AddBodyParameter(string key, string value)
             }
         }
 
+        internal void AddHttpTelemetryToHeaders(string key, string value)
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/OAuth2/OAuth2Client.cs]: Looks general purpose, maybe rename to `AddHeader`

File: tests/Microsoft.Identity.Test.Common/TestConstants.cs [hunk 2]
```diff
@@ -260,8 +260,22 @@ public static MsalTokenResponse CreateMsalTokenResponse()
...
                                                 }";
+
+        // Http Telemetry Constants
         public const string InteractiveRequestApiIdPlusCorrelationId = "1005,ad8c894a-557f-48c0-b045-c129590c344e,";
         public const string InteractiveRequestApiId = "1005";
+        public static Guid CorrelationId = new Guid("ad8c894a-557f-48c0-b045-c129590c344e");
+        public const string XClientCurrentTelemetryROPC = "2|1003,0|";
+        public const string XClientCurrentTelemetryROPCFailure = "2|1003,1003,1003,0|";
+        public const string XClientLastTelemetryROPC = "";
+        public const string XClientLastTelemetryROPCFailure =
```
Comment (bgavrilMS) [inline tests/Microsoft.Identity.Test.Common/TestConstants.cs]: I would just keep these strings inline in the tests, I doubt these are reusable.

File: src/client/Microsoft.Identity.Client/Internal/Requests/SilentRequest.cs [hunk 3]
```diff
@@ -152,6 +152,7 @@ private async Task<MsalTokenResponse> ExecuteBrokerAsync(CancellationToken cance
         private async Task<AuthenticationResult> CreateAuthenticationResultAsync(MsalAccessTokenCacheItem cachedAccessTokenItem)
         {
             var msalIdTokenItem = await CacheManager.GetIdTokenCacheItemAsync(cachedAccessTokenItem.GetIdTokenItemKey()).ConfigureAwait(false);
+            AuthenticationRequestParameters.RequestContext.ServiceBundle.TelemetryManager.SuccessfulSilentCallCount++;
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/Internal/Requests/SilentRequest.cs]: I don't think this needs to tracked separately. Doesn't "event in progress" already know about the number of silent calls made?

File: src/client/Microsoft.Identity.Client/TelemetryCore/Internal/HttpTelemeteryContent.cs [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+                return string.Empty;
+            }
+
+            IEnumerable<KeyValuePair<EventKey, EventBase>> apiEvent = eventsInProgress.Where(e => e.Key.EventName == "msal.api_event");
+
+            foreach (KeyValuePair<EventKey, EventBase> events in apiEvent)
+            {
+                events.Value.TryGetValue(MsalTelemetryBlobEventNames.ApiIdConstStrKey, out string apiId);
+                ApiId.Add(apiId);
+                events.Value.TryGetValue(MsalTelemetryBlobEventNames.ForceRefreshId, out string forceRefresh);
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/TelemetryCore/Internal/HttpTelemeteryContent.cs]: I think this constant has already been defined - see ApiTelemetryFeatureKey class.

File: src/client/Microsoft.Identity.Client/TelemetryCore/Internal/HttpTelemeteryContent.cs [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+
+        private string CreateApiIdAndCorrelationIdContent()
+        {
+            List<string> combinedApiIdCorrId = new List<string>();
+            List<string> apiId = new List<string>();
+            apiId = ApiId;
+
+            foreach (var apiIds in ApiId)
+            {
+                combinedApiIdCorrId.Add(apiIds);
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/TelemetryCore/Internal/HttpTelemeteryContent.cs]: Why do we need to combine these 2 unrelated IDs?

File: src/client/Microsoft.Identity.Client/TelemetryCore/Internal/HttpTelemeteryContent.cs [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+            // csv expected format:
+            // 2|silent_successful_count|failed_requests|errors|platform_fields
+            // failed_request can be further expanded to include:
+            // api_id_1,correlation_id_1,api_id_2,correlation_id_2|error_1,error_2
+            string data = string.Format(CultureInfo.InvariantCulture,
+                $"{TelemetryConstants.HttpTelemetrySchemaVersion2}{TelemetryConstants.HttpTelemetryPipe}{successfulSilentCallCount}{TelemetryConstants.HttpTelemetryPipe}" +
+                $"{CreateFailedRequestsContent(apiIdCorIdData)}" +
+                $"{TelemetryConstants.HttpTelemetryPipe}");
+
+            if (data.Length > 3800)
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/TelemetryCore/Internal/HttpTelemeteryContent.cs]: Please add a comment about this magic number.

File: tests/Microsoft.Identity.Test.Unit.net45/Microsoft.Identity.Test.Unit.net45.csproj [hunk 7]
```diff
@@ -14,6 +14,7 @@
       <Project>{3433eb33-114a-4db7-bc57-14f17f55da3c}</Project>
       <Name>Microsoft.Identity.Client</Name>
     </ProjectReference>
+    <ProjectReference Include="..\Microsoft.Identity.Test.Integration.net45\Microsoft.Identity.Test.Integration.net45.csproj" />
```
Comment (bgavrilMS) [inline tests/Microsoft.Identity.Test.Unit.net45/Microsoft.Identity.Test.Unit.net45.csproj]: Let's not do this :). We can share common helpers via the TestCommon assembly.

File: tests/Microsoft.Identity.Test.Unit.net45/PublicApiTests/HttpTelemetryTests2.cs [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+        public async Task AcquireToken_CreateHttpTelemetryHeaders_ReturnsSomethingInterestingAsync()
+        {
+            using (var harness = CreateTestHarness())
+            {
+                harness.HttpManager.AddInstanceDiscoveryMockHandler();
+                var factory = new HttpSnifferClientFactory();
+
+                PublicClientApplication pca = PublicClientApplicationBuilder.Create(TestConstants.ClientId)
+                                                                   .WithAuthority(new Uri(ClientApplicationBase.DefaultAuthority), true)
+                                                                   .WithTelemetry(new TraceTelemetryConfig())
```
Comment (bgavrilMS) [inline tests/Microsoft.Identity.Test.Unit.net45/PublicApiTests/HttpTelemetryTests2.cs]: Test needs to run without this.

File: tests/Microsoft.Identity.Test.Unit.net45/PublicApiTests/HttpTelemetryTests2.cs [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+        }
+
+        [TestMethod]
+        public void FetchCurrentHttpTelemetryContent_ContainsNoEventsInProgress_ReturnsEmptyString()
+        {
+            Assert.AreEqual(string.Empty, _telemetryManager.FetchCurrentHttpTelemetryContent(_trackingEvent));
+        }
+
+        [TestMethod]
+        public async Task AcquireToken_CreateHttpTelemetryHeaders_ReturnsSomethingInterestingAsync()
```
Comment (bgavrilMS) [inline tests/Microsoft.Identity.Test.Unit.net45/PublicApiTests/HttpTelemetryTests2.cs]: Can we have a more complex acceptance test ? i can help add this in - but we need to make sure that: 

- msal client exceptions are captured
- msal service exceptions are captured

My proposed acceptance test is: 

```
1.	Acquire Token Interactive successfully

Current_request = 1 | ATI_ID, 0 | 
Last_request = NULL

2.	Acquire token silent with AT served from cache
3.	Repeat step 2
4.	Acquire Token silent with force_refresh = true -> error invalid_client

Current_request = 1 | ATS_ID, 0 | 
Last_request = NULL

5.	Acquire Token silent with force_refresh = true -> error interaction_required

Current_request = 1 | ATI_ID, 0 | 
Last_request = 1 | 2 | ATS_ID, corr_step_4 | invalid_client


6.	Acquire Token interactive -> error user_cancelled (i.e. no calls to /token endpoint)
7.	Acquire Token interactive -> successful

Current_request = 1 | ATI_ID, 0 |
Last_request = 1 | 2 | | ATS_ID, corr_step_5, ATI_ID, corr_step_6 | interaction_required, user_cancelled | 

1.	Acquire Token Silent with force-refresh -> successful

Current_request = 1 | ATI_ID, 1 |
Last_request = NULL

```

File: src/client/Microsoft.Identity.Client/Internal/Requests/SilentRequest.cs (hunk 3 above)
Comment (jennyf19) [reply src/client/Microsoft.Identity.Client/Internal/Requests/SilentRequest.cs]: i couldn't find where this is tracked in the api events. lmk if you do.

---
In reply to: [389772910](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/1693#discussion_r389772910) [](ancestors = 389772910)

File: src/client/Microsoft.Identity.Client/TelemetryCore/Internal/HttpTelemeteryContent.cs (hunk 5 above)
Comment (jennyf19) [reply src/client/Microsoft.Identity.Client/TelemetryCore/Internal/HttpTelemeteryContent.cs]: the schema format for this section is: |api_id1, corr_id1, api_id2, corr_id2|

---
In reply to: [389792841](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/1693#discussion_r389792841) [](ancestors = 389792841)

File: src/client/Microsoft.Identity.Client/OAuth2/OAuth2Client.cs [hunk 10]
```diff
@@ -146,8 +151,9 @@ private bool AddCommonHeaders(RequestContext requestContext, bool addCorrelation
                 _headers.Add(OAuth2Header.AppVer, requestContext.Logger.ClientVersion);
             }
 
-            _headers.Add(TelemetryConstants.XClientLastTelemetry, _telemetryManager.FetchAndResetPreviousHttpTelemetryContent());
-            _headers.Add(TelemetryConstants.XClientCurrentTelemetry, _telemetryManager.FetchCurrentHttpTelemetryContent(requestContext.CorrelationId.AsMatsCorrelationId()));
+            //_headers.Add(TelemetryConstants.XClientLastTelemetry, _telemetryManager.FetchAndResetPreviousHttpTelemetryContent());
+            //_headers.Add(TelemetryConstants.XClientCurrentTelemetry, _telemetryManager.FetchCurrentHttpTelemetryContent());
+
```
Comment (neha-bhargava) [inline src/client/Microsoft.Identity.Client/OAuth2/OAuth2Client.cs]: Is the above code left commented for some reason?

Comment (bgavrilMS) [reply src/client/Microsoft.Identity.Client/OAuth2/OAuth2Client.cs]: This is work in progress, Jenny asked for an initial review to see if approach is valid. Apologies for not making this clear when I created a PR off her branch.

File: src/client/Microsoft.Identity.Client/OAuth2/OAuth2Client.cs (hunk 1 above)
Comment (jennyf19) [reply src/client/Microsoft.Identity.Client/OAuth2/OAuth2Client.cs]: done. htanks

---
In reply to: [389744569](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/1693#discussion_r389744569) [](ancestors = 389744569)

File: tests/Microsoft.Identity.Test.Unit.net45/Microsoft.Identity.Test.Unit.net45.csproj (hunk 7 above)
Comment (jennyf19) [reply tests/Microsoft.Identity.Test.Unit.net45/Microsoft.Identity.Test.Unit.net45.csproj]: fixed

---
In reply to: [389793604](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/1693#discussion_r389793604) [](ancestors = 389793604)

**Pull request merged**